### PR TITLE
Change matrix exponential algorithm

### PR DESCRIPTION
**Pull request by petrelharp**
I've implemented an alternative matrix exponential method as discussed in #1775 - the tests for matrix exponenetials don't pass because they're not all stochastic matrices; but something else isn't working and I haven't had a chance to look into it.

### CODE REVIEW

**Comment by petrelharp** [comment]
Ok - I think I've got a better expm algorithm in here, and the demography tests pass (at least locally?). Want to put in a test that exercises #1775, @grahamgower? I think you can push to this PR? We could merge this in, since I think this algorithm is superior, except that I've left the old expm still here for easy comparison (e.g., we'll need it to check that the new tests actually do the bad thing).

File: tests/test_demography.py [hunk 1]
```diff
@@ -1873,44 +1873,42 @@ class TestMatrixExponential:
...
     def test_identity_exp(self):
         # (-1) * np.eye(k), compared to exp(-1) * np.eye(k)
         for k in range(2, 5):
-            A = msprime.demography._matrix_exponential((-1) * np.eye(k))
-            B = np.exp(-1) * np.eye(k)
-            assert np.allclose(A, B)
+            A = (-1) * np.eye(k)
+            for i in range(k):
+                A[i, (i + 1) % k] = 1.0
+            print(A)
```
Comment (jeromekelleher) [inline tests/test_demography.py:1910]: print

**Review by jeromekelleher — COMMENTED** [review]
LGTM. To avoid having untested code in the library (pulling down the test coverage stats) we should either (a) delete it, (b) move it into the test module; (c) run some tests on it.

I think it's probably worth making a small ``test_matrix_exponential.py`` test module to isolate this stuff, since it's easy for it to get lost in the (very large) test_demography.py file.

@grahamgower, are you OK picking this one up? If you can't  push to the branch, then you can make your own PR, preserving Peter's commits?

**Comment by grahamgower** [comment]
Thanks @petrelharp, this is excellent! It works for my example from #1775, and there's a nice speed improvement too.

@jeromekelleher, yes, I can pick this up.

**Comment by jeromekelleher** [comment]
Great, thanks @grahamgower - I've pushed this into the 1.0.2 release, since you probably want a fix for this out ASAP also?

**Comment by grahamgower** [comment]
> Great, thanks @grahamgower - I've pushed this into the 1.0.2 release, since you probably want a fix for this out ASAP also?

Sounds good!

**Comment by petrelharp** [comment]
> LGTM. To avoid having untested code in the library (pulling down the test coverage stats) we should either (a) delete the old version (b) move it into the test module; (c) run some tests on it.

I think we should delete it, unless we want to move it to the test code (eg we could use it instead of scipy to compare against in validation). But I'm assuming @grahamgower has the edit token now, so I'll leave it alone.

**Pull request closed without merging**